src/backend/cpu/scan_by_key.cpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <complex>
+#include <af/dim4.hpp>
+#include <Array.hpp>
+#include <scan_by_key.hpp>
+#include <ops.hpp>
+#include <platform.hpp>
+#include <queue.hpp>
+#include <kernel/scan_by_key.hpp>
+
+using af::dim4;
+
+namespace cpu
+{
+
+    template<af_op_t op, typename Ti, typename Tk, typename To>
+    Array<To> scan(const Array<Tk>& key, const Array<Ti>& in, const int dim, bool inclusive_scan)
+    {
+        dim4 dims     = in.dims();
+        Array<To> out = createEmptyArray<To>(dims);
+        in.eval();
+
+        if (inclusive_scan) {
+            switch (in.ndims()) {
+                case 1:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 1, true> func1;
+                    getQueue().enqueue(func1, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 2:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 2, true> func2;
+                    getQueue().enqueue(func2, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 3:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 3, true> func3;
+                    getQueue().enqueue(func3, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 4:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 4, true> func4;
+                    getQueue().enqueue(func4, out, 0, key, 0, in, 0, dim);
+                    break;
+            }
+        } else {
+            switch (in.ndims()) {
+                case 1:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 1, false> func1;
+                    getQueue().enqueue(func1, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 2:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 2, false> func2;
+                    getQueue().enqueue(func2, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 3:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 3, false> func3;
+                    getQueue().enqueue(func3, out, 0, key, 0, in, 0, dim);
+                    break;
+                case 4:
+                    kernel::scan_dim_by_key<op, Ti, Tk, To, 4, false> func4;
+                    getQueue().enqueue(func4, out, 0, key, 0, in, 0, dim);
+                    break;
+            }
+        }
```

Comment:
cleaner to nest put the switch case stuff in a function and template that function from the `if` and `else`. This way it will be easier to change any parameters only in one place.